### DISCUSSION TITLE
Using ref.watch in asynchrounous calls

### DISCUSSION THREAD

**Discussion by MillieNakiganda** [rrousselGit/riverpod#3528] [Q&A]
In the riverpod documentation, there is a line that says: "The ref.watch method should not be called asynchronously". How I interpreted this was that we cannot await a ref.watch, thus not used in asynchronous methods.

But I can see that ref.watch was called in an async method in the documentation in this [example](https://riverpod.dev/docs/concepts/combining_providers#my-provider-updates-too-often-what-can-i-do) .

which is making me wonder If I got this all wrong?

**Comment by rrousselGit** [comment]
I'll reword it.

**Reply by rrousselGit** [reply]
It's fine to use ref.watch after an await inside a FutureProvider.

The point is that it shouldn't be used inside onTaps of buttons – which are event handlers called asynchronously.

**Reply by jeanlucthumm** [reply]
@rrousselGit Can we use `await ref.watch(...)` at any point in the FutureProvider or does it have to be in the beginning?

Are we allowed to `await ref.watch(...)` multiple times in a single FutureProvider? And if so in which way:

```dart
@riverpod
Future<void> example(Ref ref) async {
  final value1 = await ref.watch(prov1.future);
  final value2 = await ref.watch(prov2.future);
  // ...
}
```
```dart
@riverpod
Future<void> example(Ref ref) async {
  final [value1, value2]  = await Future.wait([
    ref.watch(prov1.future),
    ref.watch(prov2.future),
  ]);
  // ...
```
What are the implications of `prov1` changing while we're blocked on `prov2` or vice versa?
